### ISSUE TITLE
[turborepo] HEAD^ filter on shallow clone stopped working in 1.9.1

### ISSUE THREAD

**Issue by laat** [vercel/turbo#4559]
### What version of Turborepo are you using?

1.9.2

### What package manager are you using / does the bug impact?

npm

### What operating system are you using?

Linux

### Describe the Bug

Turbo tries to read a commit before HEAD^ when filtering.

`npx turbo@1.9.2 run build --filter "...[HEAD^]" --dry json` fails with the message:

```
Turbo error: failed to resolve packages to run: git error: object not found - no match for id (d9f5424696e8aa64fa7c92b88ee8d96bd529f008); class=Odb (9)
```

### Expected Behavior

It should probably work on a shallow clone

### To Reproduce

A repro, with a PR that fails with 1.9.1:

https://github.com/laat/turbo-shallow-repro/pull/2

The Github Action:
```yaml
name: Node.js CI

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
      with:
        fetch-depth: 2
    - uses: actions/setup-node@v3

    # works
    - run: npx turbo@1.9.0 run build --filter "...[HEAD^]" --dry json
    
    # fails
    - run: npx turbo@1.9.1 run build --filter "...[HEAD^]" --dry json
   
    # fails
    - run: npx turbo@1.9.2 run build --filter "...[HEAD^]" --dry json
      if: always()
```

**Comment by hugo082** [comment]
Same issue on `--filter="[HEAD~...HEAD]"`

```
ERROR  run failed: failed to resolve packages to run: git error: object not found - no match for id (9914a2c527ce7efe318d4e9f0991997066a096fc); class=Odb (9)
Turbo error: failed to resolve packages to run: git error: object not found - no match for id (9914a2c527ce7efe318d4e9f0991997066a096fc); class=Odb (9)
```

**Comment by Xa3lis** [comment]
Same issue on `--filter='[main...HEAD^1]'`

But only on one of my CI workflow
The other one is working flawlessly

```
ERROR  run failed: failed to resolve packages to run: git error: object not found - no match for id (e207f37719a34d8bb172595c541a433200d00c00); class=Odb (9)
Turbo error: failed to resolve packages to run: git error: object not found - no match for id (e207f37719a34d8bb172595c541a433200d00c00); class=Odb (9)
```

**Comment by arantespp** [comment]
Similar issue when I use git tag: `--filter=...[$TAG]`

>  ERROR  run failed: failed to resolve packages to run: git error: object is no commit object; class=Invalid (3)
Turbo error: failed to resolve packages to run: git error: object is no commit object; class=Invalid (3)

**Comment by laat** [comment]
solved in 1.9.3. 🎉 thanks @NicholasLYang

**Comment by tknickman** [comment]
Fixed in https://github.com/vercel/turbo/pull/4604

Thanks for the reports!
